_posts/2020-05-04-strangler-pattern.md

```diff
@@ -0,0 +1,115 @@
+
+Instead, you need a mechanism to refactor and replace parts of a system whilst maintaining production availability.  You require telemetry to evaluate the changes you make.
+
+# The Solution
+
+The strangler pattern provides an approach to controlled re-architecture.  
+
+It is not intended to persist indefinitely.  Instead it provides a pathway to manage internal change whilst minimising the impact to external dependencies.  Once re-architecture has been completed, a key step is the removal of the strangler from your system to simplify and move forward.
+
+![Strangler Pattern](/img/strangler-pattern/strangler-pattern.png)
+
+**Stage 1**: Identify the system boundary you want to re-architect.
+
+**Stage 2**: Create a [facade](https://en.wikipedia.org/wiki/Facade_pattern).
+
+* The facade ensures clients of the system are unaffected by the re-architecture activity and should therefore match an existing system API.
+* The API for the new service should be carefully designed to ensure backward compatibility with the legacy API.  If you want to make breaking changes to the API for the new service, that should be performed once re-architecture has been completed.
+
+**Stage 3**: Re-architect.
+
+* Implement your new service, wiring it in to the strangler facade.
+* Dark launch the new service: tee traffic between the new and legacy implementations.
+    * Return responses from the legacy implementation.  
+    * Capture, or discard, responses from the new implementation.  This enables you to monitor the behaviour of the new implementation to ensure it is fit for purpose.
+* Continuous deployment is also a good option.  Deploy partial implementations of the new service, whilst development is still in progress, to gain confidence and mitigate risk without compromising system behaviour.
+
+**Stage 4**: Cutover.
+* Once you are confident that the new implementation meets your needs, cutover production traffic so requests and responses are served from the new implementation.
+* Retain the legacy implementation to allow for easy rollback if required.
+
+**Stage 5**: Clean up the mess.
+* Once the new system is established delete the legacy code that has been replaced.
+
+**Stage 6**: Migrate clients to the new API and remove the strangler facade
+* This is likely to require breaking changes to clients, and should be managed accordingly.
+* To manage the transition, if you have a large number of clients, expose the new API whilst maintaining the strangler facade.  
+    * Monitor usage of the strangler facade and proactively engage with clients to help them migrate.  
+    * Once the strangler facade is no longer being used, it can be safely removed.
+
+# Considerations
+
+* Break up the system based on new service boundaries or integration layers.
```

Comment:
can we link for "service boundaries" and "integration layers" ? 
Martin Fowler must have something surely.